### PR TITLE
[wmcb] add service dependency and service dependent restart

### PR DESCRIPTION
**Pull request by mansikulkarni96**
This PR adds service dependent stopping and starting in Windows services. It also adds a dependency from kubelet on underlying docker service.

### CODE REVIEW

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [inline pkg/bootstrapper/bootstrapper.go:120]: Missing doc

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [inline pkg/bootstrapper/bootstrapper.go:642]: Instead of returning `bool` and `error` should we just return `error` if the service is not running and `nil` otherwise?

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [inline pkg/bootstrapper/bootstrapper.go:655]: Why are we not capturing the error here ?

File: pkg/bootstrapper/bootstrapper.go
Comment (VaishnaviHire) [inline pkg/bootstrapper/bootstrapper.go:193]: Why are we not capturing the error here ?

**Review by VaishnaviHire — COMMENTED** [review]
Thanks @mansikulkarni96 for working on this. PTAL at my comments

File: pkg/bootstrapper/bootstrapper.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go:642]: the way we use this method is we check for the bool value and decide if there is anything to be done in case of stopping and starting a service, if the service is already in expected state we return nil. eg: If the service is not running and we want to stop it, we could just return nil as the variable value is false(returning an error because its not running doesn't fit here)

File: pkg/bootstrapper/bootstrapper.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go:193]: the reason is (// If there is already a kubelet service running, find it and assign it) if there isn't a kubelet service already present we don't need to make sure it is assigned here. It will get created in createService and be assigned there.

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:623]:
```suggestion
	if len(services) == 0 {
	    return nil, nil
	}
	// rest of the code here with less indentation
```

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:687]: For each service you find, you should be instantiating a new service object and calling `start` on it. This again looks a bit odd to me otherwise.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this PR, @mansikulkarni96. Please address my comments.

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [reply pkg/bootstrapper/bootstrapper.go:193]: I think we should be careful about doing this, what if there is a kubelet service running and we cannot open it for whatever reason? (permissions or something).
We would need to capture that error. 

If setService is being used in this way I think it should either be fed a service object, or we should check that the service exists before calling it, then only call it if the service does exist, and error out if it returns an error

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [inline pkg/bootstrapper/bootstrapper.go:200]: nit: would be more descriptive to say `service manager is nil`

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [inline pkg/bootstrapper/bootstrapper.go:120]: It seems like this holds all the service specific information strictly for the kubelet service

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [inline pkg/bootstrapper/bootstrapper.go:530]: This function confuses me. It says it `starts a Windows service`, but doesn't give input for which service.
I feel like this should be a method of the `service` object. It can then be renamed to `start()`

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [inline pkg/bootstrapper/bootstrapper.go:551]: Same as the comment on `startService`

File: pkg/bootstrapper/bootstrapper.go
Comment (sebsoto) [inline pkg/bootstrapper/bootstrapper.go:140]: This struct is written like it is meant for a generic service, however it includes the service manager within it. If we had multiple services, we end up with multiple service managers. IMO this struct should either become specific to the kubelet service, or have the manager moved out of it.

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/bootstrapper/bootstrapper.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go]: I am capturing the error for setService(), it is not possible to do it in this particular call as the service is initialized here and service object will always be nil, hence we cannot validate it here.

File: pkg/bootstrapper/bootstrapper.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go]: I am calling the stop service recursively hence I am creating a new list here which gets updated on each call, also added some more comments around it.

**Review by LorbusChris — APPROVED** [review]

File: pkg/bootstrapper/service.go
Comment (aravindhp) [inline pkg/bootstrapper/service.go:120]: `kubelet's dependent service`

Please call out that you are only going one level deep. i.e. you are not recursively checking the dependencies.

File: pkg/bootstrapper/service.go
Comment (aravindhp) [inline pkg/bootstrapper/service.go:69]: Same trick mentioned before can be done here:
```suggestion
	if len(service.svcDependents) == 0 {
	    return nil
	}
	// rest of the code with less indentation
```

File: pkg/bootstrapper/service.go
Comment (aravindhp) [inline pkg/bootstrapper/service.go:126]: Is it possible to link to the example itself?

File: pkg/bootstrapper/service.go
Comment (ravisantoshgudimetla) [inline pkg/bootstrapper/service.go:12]: Technically you specialized here but the fields are generic enough. However there is no problem with this approach considering we just have a single service for now.

File: pkg/bootstrapper/service.go
Comment (aravindhp) [inline pkg/bootstrapper/service.go:127]: Is it? There is definitely a time you are waiting here. Only `serviceWaitTime` is arbitrary.

File: pkg/bootstrapper/service.go
Comment (sebsoto) [inline pkg/bootstrapper/service.go:232]: extra newline

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @mansikulkarni96. Please address my comments. I am leaving the struct design review to @ravisantoshgudimetla and @sebsoto.

File: pkg/bootstrapper/service.go
Comment (sebsoto) [inline pkg/bootstrapper/service.go:217]: This mentions kubelet but seems generic

File: pkg/bootstrapper/service.go
Comment (sebsoto) [inline pkg/bootstrapper/service.go:226]: I think this should be renamed to `stopAndRemove()` the service is implicit as it is a method of the kubeletService struct

File: pkg/bootstrapper/service.go
Comment (sebsoto) [inline pkg/bootstrapper/service.go:208]: This can be renamed to `isRunning()` as it is a method of kubeletService

File: pkg/bootstrapper/service.go
Comment (sebsoto) [inline pkg/bootstrapper/service.go:200]: Can be renamed to remove() as it is a method of kubelet service

File: pkg/bootstrapper/service.go
Comment (sebsoto) [inline pkg/bootstrapper/service.go:183]: Can be renamed to refresh() as it is a method of kubelet service

File: pkg/bootstrapper/service.go
Comment (sebsoto) [inline pkg/bootstrapper/service.go:168]: Mentions kubelet service but this seems generic.

in addition should this be renamed to `stopDependantServices`? As there could be more than one dependant service

File: pkg/bootstrapper/service.go
Comment (sebsoto) [reply pkg/bootstrapper/service.go:217]: Does this function actually check the dependants, or does it just check the status of whats passed into it?
If the latter, this can be renamed to `isRunning`.

File: pkg/bootstrapper/service.go
Comment (sebsoto) [inline pkg/bootstrapper/service.go:143]: Can be renamed to stop() as it is a method of kubelet service

File: pkg/bootstrapper/service.go
Comment (sebsoto) [inline pkg/bootstrapper/service.go:121]: Mentions kubelet but seems generic

File: pkg/bootstrapper/service.go
Comment (sebsoto) [inline pkg/bootstrapper/service.go:81]: Does this function actually start the dependants, or does it just start the service of whats passed into it?
If the latter, this should be renamed to startService. A function should be named according to what it does, not how its intended to be used

File: pkg/bootstrapper/service.go
Comment (sebsoto) [inline pkg/bootstrapper/service.go:53]: can be renamed to `start()` as its a method of kubeletService

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: pkg/bootstrapper/kubelet.go
Comment (mansikulkarni96) [inline pkg/bootstrapper/kubelet.go:98]: mentioning the approach here is not recursive/goes one level deep

File: pkg/bootstrapper/bootstrapper.go
Comment (mansikulkarni96) [inline pkg/bootstrapper/bootstrapper.go:47]: timeout required now to stop two services and refresh.

Comment (aravindhp) [reply pkg/bootstrapper/bootstrapper.go:47]: Please update the comment to reflect that.

File: pkg/bootstrapper/kubelet.go
Comment (aravindhp) [reply pkg/bootstrapper/kubelet.go:98]: Please mention it in the `stop()` docs too. A reader of the docs should be able to see that.

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:495]: Is this service going to be named `docker` across cloud providers / Windows images?

File: pkg/bootstrapper/kubelet.go
Comment (aravindhp) [inline pkg/bootstrapper/kubelet.go:31]: Sorry I should have caught this before. Shouldn't this be just `Config()` as per go naming conventions?

File: pkg/bootstrapper/kubelet.go
Comment (aravindhp) [inline pkg/bootstrapper/kubelet.go:62]: Please mention the dependent service in the error message.

File: pkg/bootstrapper/kubelet.go
Comment (aravindhp) [inline pkg/bootstrapper/kubelet.go:211]: Please mention what that arbitrary wait time is.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: pkg/bootstrapper/bootstrapper.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go:47]: done

File: pkg/bootstrapper/bootstrapper.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go:495]: The name docker is independent of cloud providers, the package management msft used to install docker package on any Windows server images, installs them with service named "docker"

File: pkg/bootstrapper/kubelet.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/kubelet.go]: done

File: pkg/bootstrapper/kubelet.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/kubelet.go]: done

File: pkg/bootstrapper/kubelet.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/kubelet.go]: done

File: pkg/bootstrapper/kubelet.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/kubelet.go:99]: done

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go]: Nit: this formatting seems inconsistent

**Review by aravindhp — COMMENTED** [review]
LGTM from me. I will leave it to @ravisantoshgudimetla to approve and @sebsoto to officially LGTM.

**Comment by sebsoto** [comment]
Thanks for making the changes @mansikulkarni96 i will LGTM once this is approved

File: pkg/bootstrapper/bootstrapper.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go:494]: this change is a result of adding a dependency param, if the formatting is left untouched verify job fails, hence had to update it.

File: pkg/bootstrapper/kubelet.go
Comment (ravisantoshgudimetla) [inline pkg/bootstrapper/kubelet.go:14]: I thought you wouldn't need this field, you could have directly embedded the mgr.Service and that would have given the necessary options?

**Review by ravisantoshgudimetla — COMMENTED** [review]
Mostly LGTM. I just have question around why the embedding didn't work.

File: pkg/bootstrapper/kubelet.go
Comment (mansikulkarni96) [reply pkg/bootstrapper/kubelet.go:14]: that is right, but I saw no harm in keeping this field to access the service object as I had already implemented using it.

File: pkg/bootstrapper/bootstrapper.go
Comment (aravindhp) [reply pkg/bootstrapper/bootstrapper.go:494]: Yeah, adding the comment has caused things to be inconsistent. But it is fine to leave it as is. This is the case to add comments on the same line.

**Review by ravisantoshgudimetla — APPROVED** [review]
/approve

Thank you for working on this @mansikulkarni96

**Pull request merged**
